Backports for Julia 1.10.8

KristofferC opened the pull request:
Backported PRs:
- [x] #55773 <!-- Add compat entry for `Base.donotdelete` -->
- [x] #56311 <!-- Fix `log_quasitriu` for internal scaling `s=0` -->
- [x] #56640 <!-- subtype: fast path for Type == TypeVar -->
- [x] #56784 <!-- fix `exp(weirdNaN)` -->
- [x] #56787 <!-- Fix partially_inline for unreachable -->
- [x] #56411 <!-- xref `UnionAll` in the doc string of `where` -->
- [x] #56772 <!-- docs: fix edge case in rational number conversion `float(a//b)` -->
- [x] #56842 <!-- dict docs: document that ordering of keys/values/pairs match iterate -->
- [x] #56793 <!-- Extend `Base.rationalize` instead of defining new function `Base.MathConstants.rationalize` -->
- [x] #43845 <!-- Add preference for version named manifest files -->

Need manual backport:
- [ ] #55405 <!-- handle unbound vars in NTuple fields -->
- [ ] #55744 <!-- fix #45494, error in ssa conversion with complex type decl -->
- [ ] #56789 <!-- Fix test report alignment -->
- [ ] #56801 <!-- gc: improve mallocarrays locality -->

Contains multiple commits, manual intervention needed:
- [x] #54738 <!-- serialization: fix relocatability bug -->
- [ ] #55348 <!-- Materialize complex Symmetric matrices in eigen -->
- [ ] #55306 <!-- AllocOpt: Fix stack lowering where alloca continas boxed and unboxed data -->
- [ ] #55729 <!-- Avoid confounding compilation side effects of `@time_imports` -->
- [ ] #55877 <!-- fix FileWatching designs and add workaround for a stat bug on Apple -->
- [ ] #56369 <!-- cache the `find_all_in_cache_path` call during parallel precompilation -->
- [ ] #56755 <!-- docs: fix scope type of a `struct` to hard -->
- [x] #56914 <!-- Don't report only-inferred methods as recompiles -->

Non-merged PRs with backport label:
- [ ] #56915 <!-- Compiler: fix `tmerge(Const(s), Const(t))` st. `(s !== t) && (s == t)` -->
- [ ] #56890 <!-- Enable getting non-boxed LLVM type from Julia Type -->
- [ ] #56813 <!-- Miscellaneous `Meta.partially_inline!` fixes -->
- [ ] #55958 <!-- also redirect JL_STDERR etc. when redirecting to devnull -->
- [ ] #55956 <!-- Make threadcall gc safe -->
- [ ] #55813 <!-- Check for conflicting `@ccallable` name before JIT registration -->
- [ ] #55534 <!-- Set stdlib sources as read-only during installation -->
- [ ] #55499 <!-- propagate the terminal's `displaysize` to the `IOContext` used by the REPL -->
- [ ] #55311 <!-- docs: Add headers for easier Core, Base and stdlib reference -->
- [ ] #55220 <!-- `isfile_casesensitive` fixes on Windows -->
- [ ] #55169 <!-- `propertynames` for SVD respects private argument -->
- [ ] #51479 <!-- prevent code loading from lookin in the versioned environment when building Julia -->
- [ ] #50813 <!-- More doctests for Sockets and capitalization fix -->
- [ ] #50157 <!-- improve docs for `@inbounds` and `Base.@propagate_inbounds` -->

topolarity committed:
Handle no-postdominator case in finalizer pass

This pass was assuming that the post-dominator of all finalizer uses
exists as a real BB in the CFG.

jishnub committed:
Restrict binary ops for Diagonal and Symmetric to Number eltypes (#55251)

The `(::Diagonal) + (::Symmetric)` and analogous methods were
specialized in https://github.com/JuliaLang/julia/pull/35333 to return a
`Symmetric`, but these only work if the `Diagonal` is also symmetric.
This typically holds for arrays of numbers, but may not hold for
block-diagonal and other types for which symmetry isn't guaranteed. This
PR restricts the methods to arrays of `Number`s.

Fixes, e.g.:
```julia
julia> using StaticArrays, LinearAlgebra

julia> D = Diagonal(fill(SMatrix{2,2}(1:4), 2))
2×2 Diagonal{SMatrix{2, 2, Int64, 4}, Vector{SMatrix{2, 2, Int64, 4}}}:
 [1 3; 2 4]      ⋅
     ⋅       [1 3; 2 4]

julia> S = Symmetric(D)
2×2 Symmetric{AbstractMatrix, Diagonal{SMatrix{2, 2, Int64, 4}, Vector{SMatrix{2, 2, Int64, 4}}}}:
 [1 3; 3 4]      ⋅
     ⋅       [1 3; 3 4]

julia> S + D
2×2 Symmetric{AbstractMatrix, Diagonal{SMatrix{2, 2, Int64, 4}, Vector{SMatrix{2, 2, Int64, 4}}}}:
 [2 6; 6 8]      ⋅
     ⋅       [2 6; 6 8]

julia> S[1,1] + D[1,1]
2×2 SMatrix{2, 2, Int64, 4} with indices SOneTo(2)×SOneTo(2):
 2  6
 5  8

julia> (S + D)[1,1] == S[1,1] + D[1,1]
false
```
After this,
```julia
julia> S + D
2×2 Matrix{AbstractMatrix{Int64}}:
 [2 6; 5 8]  [0 0; 0 0]
 [0 0; 0 0]  [2 6; 5 8]
```

Even with `Number`s as elements, there might be an issue with `NaN`s
along the diagonal as `!issymmetric(NaN)`, but that may be a different
PR.

(cherry picked from commit 197295c84aab217bffde01a4339f1d0e8e95fb98)

LilithHafner committed:
Add compat entry for `Base.donotdelete` (#55773)

(cherry picked from commit 346f38bceabf3dab1d3912fe822a663735c91d4a)

aravindh-krishnamoorthy committed:
Fix `log_quasitriu` for internal scaling `s=0` (#56311)

This PR is a potential fix for #54833.

## Description
The function
https://github.com/JuliaLang/julia/blob/2a06376c18afd7ec875335070743dcebcd85dee7/stdlib/LinearAlgebra/src/triangular.jl#L2220
computes $\boldsymbol{A}^{\dfrac{1}{2^s}} - \boldsymbol{I}$ for a
real-valued $2\times 2$ matrix $\boldsymbol{A}$ using Algorithm 5.1 in
[R1]. However, the algorithm in [R1] as well as the above function do
not handle the case $s=0.$ This fix extends the function to compute
$\boldsymbol{A}^{\dfrac{1}{2^s}} - \boldsymbol{I} \Bigg|_{s=0} =
\boldsymbol{A} - \boldsymbol{I}.$

## Checklist
- [X] Fix code: `stdlib\LinearAlgebra\src\triangular.jl` in function
`_sqrt_pow_diag_block_2x2!(A, A0, s)`.
- [X] Add test case: `stdlib\LinearAlgebra\test\triangular.jl`.
- [X] Update `NEWS.md`.
- [X] Testing and self review.

|  Tag  | Reference |
| --- | --- |
| <nobr>[R1]</nobr> | Al-Mohy, Awad H. and Higham, Nicholas J. "Improved
Inverse Scaling and Squaring Algorithms for the Matrix Logarithm", 2011,
url: https://eprints.maths.manchester.ac.uk/1687/1/paper11.pdf |

---------

Co-authored-by: Daniel Karrasch <daniel.karrasch@posteo.de>
Co-authored-by: Oscar Smith <oscardssmith@gmail.com>
(cherry picked from commit 2cdfe062952c3a1168da7545a10bfa0ec205b4db)

christiangnrd commented:
https://github.com/JuliaLang/Pkg.jl/pull/4093

topolarity commented:
> [JuliaLang/Pkg.jl#4093](https://github.com/JuliaLang/Pkg.jl/pull/4093)

And a follow-up to that backport: https://github.com/JuliaLang/julia/pull/56666

DilumAluthgeBot committed:
🤖 [backports-release-1.10] Bump the Pkg stdlib from 06f7a7e5b to 845909f49 (#56669)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>
fix preferential Manifest.toml naming (#3731)

topolarity committed:
Prevent pre-compilation package from triggering its own extensions (#56666)

It is possible for an extension `ExtAB` to be loadable by one of its
triggers, e.g. if `A` loads `B`. However, this loading is not supposed
to happen during pre-compilation of `A`.

Getting this wrong means disagreeing with the scheduled pre-compile jobs
(`A` is not scheduled to depend on or generate a cache file for `ExtAB`
but accidentally attempts both) and leads to confusing errors about
missing cache files.

We used to cover up this bad behavior w/ an erroneous cycle warning
(fixed by #55910), but now we need to be sure this works.

topolarity commented:
Backport of #55251 looks like it has a bad test: https://buildkite.com/julialang/julia-release-1-dot-10/builds/433#01935fa6-46db-4194-b0dc-27b410bd4db8/867-1134

N5N3 committed:
subtype: fast path for Type == TypeVar (#56640)

close #56606

(cherry picked from commit 7df1dfa2f1a51755eb35a2b5e51c810674dc10ee)

IanButterworth committed:
Add preference for version named manifest files (#43845) (#56600)

KristofferC committed:
bump SparseArrays to latest v1.10

aviatesk committed:
update irutils.jl and fix the irpasses.jl test

oscardssmith committed:
fix `exp(weirdNaN)` (#56784)

Fixes https://github.com/JuliaLang/julia/issues/56782
Fix `exp(reinterpret(Float64, 0x7ffbb14880000000))` returning non-NaN
value. This should have minimal performance impact since it's already in
the fallback branch.

(cherry picked from commit 19e06d39141e1e2f0f9c4664ba9b88867b587a6e)

wsmoses committed:
Fix partially_inline for unreachable (#56787)

(cherry picked from commit 9118ea7565feae13d5b47654a3c245c0df36e753)

Commit:
bump Pkg to latest 1.10

nsajko committed:
xref `UnionAll` in the doc string of `where` (#56411)

(cherry picked from commit 8924e6f6c4f61946e72fde783c59a7996d749051)

Commit:
docs: fix edge case in rational number conversion `float(a//b)` (#56772)

Fixes #56726 added the changes that were suggested. fixing the mistake.

---------

Co-authored-by: Max Horn <max@quendi.de>
Co-authored-by: Chengyu Han <git@wo-class.cn>
(cherry picked from commit fc9e7c4a6a36d7d54ae7dd8b89ac14a8334a59c4)

cossio committed:
dict docs: document that ordering of keys/values/pairs match iterate (#56842)

Fix #56841.

Currently the documentation states that keys(dict) and values(dict)
iterate in the same order. But it is not stated whether this is the same
order as that used by pairs(dict), or when looping, for (k,v) in dict.

This PR makes this guarantee explicit.

(cherry picked from commit 796d823b4fbb3ec94c09f7022f9b614fc846dced)

sostock committed:
Extend `Base.rationalize` instead of defining new function (#56793)

#55886 accidentally created a new function
`Base.MathConstants.rationalize` instead of extending
`Base.rationalize`, which is the reason why `Base.rationalize(Int, π)`
isn’t constant-folded in Julia 1.10 and 1.11:

```
julia> @btime rationalize(Int,π);
  1.837 ns (0 allocations: 0 bytes)      # v1.9: constant-folded
  88.416 μs (412 allocations: 15.00 KiB) # v1.10: not constant-folded
```

This PR fixes that. It should probably be backported to 1.10 and 1.11.

(cherry picked from commit d3c26b7ecce7e8319788bc0a035241d6dfd21ac0)

christiangnrd commented:
Anyone else on macOS getting errors trying to load Distributed with this branch? Doesn't seem to happen on linux nor 1.10.7, 1.11.2, and master (latest nightly cab11bbd2d5).

```
ChrichriMBP:~/julia$ j +dev
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.6 (2024-12-03)
 _/ |\__'_|_|_|\__'_|  |  backports-release-1.10/53ba326395* (fork: 503 commits, 540 days)
|__/                   |

julia> using Distributed
ERROR: ArgumentError: Package Distributed not found in current path.
- Run `import Pkg; Pkg.add("Distributed")` to install the Distributed package.
Stacktrace:
 [1] macro expansion
   @ ./loading.jl:1842 [inlined]
 [2] macro expansion
   @ ./lock.jl:267 [inlined]
 [3] __require(into::Module, mod::Symbol)
   @ Base ./loading.jl:1823
 [4] #invoke_in_world#3
   @ ./essentials.jl:926 [inlined]
 [5] invoke_in_world
   @ ./essentials.jl:923 [inlined]
 [6] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:1816
```

IanButterworth commented:
Looks like your branch is old. Says 1.10.6

christiangnrd commented:
I think this branch was created before the commit to change the version to 1.10.7 was merged https://github.com/JuliaLang/julia/blob/39cfea14d5224138e980ab7a7a0c551325e249e8/VERSION

Commit:
Revert "Restrict binary ops for Diagonal and Symmetric to Number eltypes (#55251)"

This reverts commit 30573bee130b7d55ce6e1ef3a595b4aa12be5d1a.

IanButterworth committed:
Don't report only-inferred methods as recompiles (#56914)

(cherry picked from commit 6e28217cb6e1b9dec9669c6c4a44d3c6cf694ca2)

IanButterworth committed:
REPL: Limit method lookup when completing kwargs (#56963)

(cherry picked from commit 10e20f12679a549d82337af14a2caa709e03b88b)

giordano committed:
[BinaryPlatforms] Parse `rc64`/`riscv64` triplets

This is partial backport of #56105, only the part relative to parsing triplets.

giordano commented:
I pushed 96ad0a8c09eb35286c553646b341454b84544395 to do the same thing as described in https://github.com/JuliaLang/julia/pull/56741#issuecomment-2577773634

KristofferC commented:
@wsmoses, @gbaraldi, it seems the SparseArray bump (with the hvcat change) breaks the doctests?

<img width="1084" alt="image" src="https://github.com/user-attachments/assets/ac5f3b5b-9fb8-47f0-a427-24e8a70dbd95" />

wsmoses commented:
@KristofferC I'm not quite sure how to do the baked in stdlib, but @gbaraldi tried https://github.com/JuliaSparse/SparseArrays.jl/pull/592 and it seemed to resolve ?

DilumAluthgeBot committed:
🤖 [backports-release-1.10] Bump the SparseArrays stdlib from 78035e1 to 248d9f9 (#57005)

Stdlib: SparseArrays
URL: https://github.com/JuliaSparse/SparseArrays.jl.git
Stdlib branch: release-1.10
Julia branch: backports-release-1.10
Old commit: 78035e1
New commit: 248d9f9
Julia version: 1.10.7
SparseArrays version: 1.10.0(Does not match)
Bump invoked by: @dkarrasch
Powered by:
[BumpStdlibs.jl](https://github.com/JuliaLang/BumpStdlibs.jl)

Diff:
https://github.com/JuliaSparse/SparseArrays.jl/compare/78035e1a4475b77f54964b3f9dd7ca6b92ac7f51...248d9f94cf31d9668e5b07f237f13a239472b2ec

```
$ git log --oneline 78035e1..248d9f9
248d9f9 More adhereance to 1.10 error types (hopefully) (#592)
```

Co-authored-by: dkarrasch <26658441+dkarrasch@users.noreply.github.com>

DilumAluthgeBot committed:
🤖 [backports-release-1.10] Bump the Pkg stdlib from 0ac49eb3e to 6390ea92b (#57014)

Co-authored-by: IanButterworth <1694067+IanButterworth@users.noreply.github.com>

vchuravy committed:
Backport "serialization: fix relocatability bug" to 1.10 (#56973)

(cherry picked from commit cf4f1bab961d0e99e8cfdd4c3ef0acb3bd3767fb)

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

KristofferC commented:
@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/56653#issuecomment-2584651211) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c9b78ce_vs_4976d05/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**4 packages crashed only on the current version.**

 - The process was aborted: 4 packages

206 packages crashed on the previous version too.

## ✖ Packages that failed

**25 packages failed only on the current version.**

 - Illegal method overwrites during precompilation: 2 packages
 - Package has test failures: 5 packages
 - Package tests unexpectedly errored: 2 packages
 - Networking-related issues were detected: 1 packages
 - Tests became inactive: 3 packages
 - Test duration exceeded the time limit: 12 packages

2402 packages failed on the previous version too.

## ✔ Packages that passed tests

**36 packages passed tests only on the current version.**

 - Other: 36 packages

6423 packages passed tests on the previous version too.

## ➖ Packages that were skipped altogether

1082 packages were skipped on the previous version too.



</details>

IanButterworth committed:
disable flaky tests - part backport from #53682

IanButterworth commented:
The tests that are failing on aarch64-macos were turned into skips in https://github.com/JuliaLang/julia/pull/53682 see https://github.com/JuliaLang/julia/pull/53682#issuecomment-1994040478 so I backported that part here

IanButterworth commented:
The remaining issue is on `test i686-linux-gnu`, and I think first appeared on [`c9b78ce` (#56653)](https://github.com/JuliaLang/julia/pull/56653/commits/c9b78ce0333006e229830172b9152f6c22a0bca9)
```
  | 2025-01-11 22:27:25 EST | Executing tests that run on node 1 only:
  | 2025-01-11 22:27:26 EST | ccall                                             (1) \|        started at 2025-01-12T03:27:26.137
  | 2025-01-11 22:28:05 EST | fatal: error thrown and no exception handler available.
  | 2025-01-11 22:28:05 EST | ErrorException("fatal error allocating signal stack: mmap: Cannot allocate memory")
  | 2025-01-11 22:28:05 EST | ijl_errorf at /cache/build/builder-amdci4-3/julialang/julia-release-1-dot-10/src/rtutils.c:77
  | 2025-01-11 22:28:05 EST | alloc_sigstack at /cache/build/builder-amdci4-3/julialang/julia-release-1-dot-10/src/signals-unix.c:636 [inlined]
  | 2025-01-11 22:28:05 EST | jl_install_thread_signal_handler at /cache/build/builder-amdci4-3/julialang/julia-release-1-dot-10/src/signals-unix.c:643
  | 2025-01-11 22:28:05 EST | jl_init_root_task at /cache/build/builder-amdci4-3/julialang/julia-release-1-dot-10/src/task.c:1742
  | 2025-01-11 22:28:05 EST | ijl_adopt_thread at /cache/build/builder-amdci4-3/julialang/julia-release-1-dot-10/src/threading.c:429
  | 2025-01-11 22:28:05 EST | unknown function (ip: 0xe744c80b)
  | 2025-01-11 22:28:05 EST | jl_work_wrapper at /cache/build/builder-amdci4-3/julialang/julia-release-1-dot-10/src/jl_uv.c:1051
  | 2025-01-11 22:28:05 EST | uv__queue_work at /workspace/srcdir/libuv/src/threadpool.c:305
  | 2025-01-11 22:28:05 EST | worker at /workspace/srcdir/libuv/src/threadpool.c:122
  | 2025-01-11 22:28:05 EST | start_thread at /lib/i386-linux-gnu/libpthread.so.0 (unknown line)
  | 2025-01-11 22:28:05 EST | clone at /lib/i386-linux-gnu/libc.so.6 (unknown line)
```

KristofferC commented:
Seeing

```
Failed to precompile TimeseriesSurrogatesVisualizations [bde4fa71-810a-59c6-83e0-28def334d148] to "/home/pkgeval/.julia/compiled/v1.10/TimeseriesSurrogatesVisualizations/jl_rQfE4B".
julia: /source/src/subtype.c:4669: sub_msp: Assertion `obvious_sub == 3 || obvious_sub == subtype || ijl_has_free_typevars(x) || ijl_has_free_typevars(y)' failed.

[82] signal (6.-6): Aborted
in expression starting at /home/pkgeval/.julia/packages/Makie/Y3ABD/precompile/shared-precompile.jl:90
unknown function (ip: 0x756edb53cebc)
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x756edb4d8394)
__assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
sub_msp at /source/src/subtype.c:4669
type_morespecific_ at /source/src/subtype.c:4927
type_morespecific_ at /source/src/subtype.c:5092
type_morespecific_ at /source/src/subtype.c:5092
type_morespecific_ at /source/src/subtype.c:5037
type_morespecific_ at /source/src/subtype.c:4995
type_morespecific_ at /source/src/subtype.c:5092
type_morespecific_ at /source/src/subtype.c:5089
```

and

```
Failed to precompile CairoMakie [13f3f980-e62b-5c42-98c6-ff1f3baf88f0] to "/home/pkgeval/.julia/compiled/v1.10/CairoMakie/jl_efx9pP".
julia: /source/src/subtype.c:4669: sub_msp: Assertion `obvious_sub == 3 || obvious_sub == subtype || ijl_has_free_typevars(x) || ijl_has_free_typevars(y)' failed.

[187] signal (6.-6): Aborted
in expression starting at /home/pkgeval/.julia/packages/Makie/Y3ABD/precompile/shared-precompile.jl:90
unknown function (ip: 0x7672338a1ebc)
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x76723383d394)
__assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
sub_msp at /source/src/subtype.c:4669
type_morespecific_ at /source/src/subtype.c:4927
type_morespecific_ at /source/src/subtype.c:5092
type_morespecific_ at /source/src/subtype.c:5092
type_morespecific_ at /source/src/subtype.c:5037
type_morespecific_ at /source/src/subtype.c:4995
type_morespecific_ at /source/src/subtype.c:5092
```

KristofferC commented:
But also reproduces on 1.10.7, these are actually there in the PkgEval log as well, just that they get recorded as a test failure from later.

KristofferC commented:
JuliaInterpreter failure is due to a check

```
if !Base.GIT_VERSION_INFO.tagged_commit # only run on nightly
```

which is not a correct way to run on nightly.

KristofferC committed:
fix kwarg limit method backport
